plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XProjectManager.java

```diff
@@ -197,6 +197,10 @@ protected XBuildResult doBuild(Set<URI> dirtyFiles, Set<URI> deletedFiles,
 
 		XBuildResult result = incrementalBuilder.build(request);
 
+		if (!request.isGeneratorEnabled()) {
+			workspaceManager.addFilesAwaitingGeneration(result.getAffectedResources());
```

Comment:
I guess you mean `XWorkspaceManager#getIncrementalDirtyBuildable()` instead of `XWorkspaceManager#getIncrementalGenerateBuildable()`, right?

Reply:
True, that looks like a *much* better location. The entire handling of "stuff awaiting generation" will then be local to a single class (at least for now), and also one of @szarnekow 's comments will be resolved by that.

However, this means we have to be sure that in case of cancellation we will receive a partial result instead of exiting with a CancellationException. So far, neither the API docs (e.g. that of `XBuildable#build(CancelIndicator)`) nor the actual code (e.g. [1]) looks as if that is guaranteed. But I'll have a look.

[1] https://github.com/eclipse/n4js/blob/514cf619820ab9c67f1e57313e475d380f8af2a1/plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XBuildManager.java#L273-L274